src/navigation/_CollapsingNavigationMenu.jsx

```diff
@@ -0,0 +1,162 @@
+
+import styles from './CollapsingNavigationMenu.module.scss';
+
+const cx = classNames.bind(styles);
+
+const propTypes = {
+  /**
+   * Currently selected path
+   */
+  selectedPath: PropTypes.string,
+
+  /**
+   * menu items to display.
+   */
+  menuItems: menuItemPropType.isRequired,
+
+  /**
+   * On select callback
+   */
+  onSelect: PropTypes.func.isRequired,
+};
+
+const defaultProps = {
+  selectedPath: undefined,
+};
+
+class CollapsingNavigationMenu extends React.Component {
+  /**
+   * Returns a list of keys in the tree to 'open' that lead to the selected path.
+   * @param {*} item the item to traverse
+   * @param {*} selectedPath the currently selected path
+   */
+  static keysToItem(item, selectedPath) {
+    let paths = [];
+    if (item.childItems) {
+      item.childItems.some((childItem) => {
+        if (selectedPath === childItem.path) {
+          paths = [item.path];
+          // if found bail early.
+          return true;
+        }
+
+        const childPaths = CollapsingNavigationMenu.keysToItem(childItem, selectedPath);
+        if (childPaths.length > 0) {
+          paths = childPaths.concat([item.path]);
+          // if found bail early.
+          return true;
+        }
+        return false;
+      });
+    }
+
+    return paths;
+  }
+
+  /**
+   * Returns an object containing the keys of the items to open to reveal the selected path in the tree.
+   * @param {*} menuItems list of all menu items
+   * @param {*} selectedPath the currently selected path
+   */
+  static openKeysToItem(menuItems, selectedPath) {
+    return CollapsingNavigationMenu.keysToItem(menuItems, selectedPath).reduce((acc, path) => {
+      acc[path] = true;
+      return acc;
+    }, {});
+  }
+
+  static getDerivedStateFromProps({ menuItems, selectedPath }, state) {
+    const newState = {};
+    if (state.previousSelectedPath !== selectedPath) {
+      newState.openKeys = Object.assign({}, state.openKeys, CollapsingNavigationMenu.openKeysToItem(menuItems[0], selectedPath));
+      newState.previousSelectedPath = selectedPath;
+    }
+    return newState;
+  }
+
+  constructor(props) {
+    super(props);
+    const { menuItems, selectedPath } = props;
+    this.renderMenuItems = this.renderMenuItems.bind(this);
+    this.handleOnClick = this.handleOnClick.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+
+    this.state = {
+      previousSelectedPath: selectedPath,
+      openKeys: CollapsingNavigationMenu.openKeysToItem(menuItems[0], selectedPath),
+    };
+  }
+
+  handleKeyDown(event, item) {
+    if (event.nativeEvent.keyCode === KeyCode.KEY_SPACE || event.nativeEvent.keyCode === KeyCode.KEY_RETURN) {
+      this.handleOnClick(event, item);
+    }
+  }
+
+  handleOnClick(event, item) {
+    const { onSelect } = this.props;
+    const { openKeys } = this.state;
+
+    if (!item.childItems) {
+      onSelect(item.path);
+      return;
+    }
+    openKeys[item.path] = !openKeys[item.path];
```

Comment:
What's this doing?

Reply:
Yeah, exactly that. It toggles it to be the opposite of the current setting.